### ISSUE TITLE
2FA module asks for authentication no matter the credentials provided

### ISSUE THREAD

**Issue by hpenney2** [DSJAS/DSJAS#25]
When the two-factor authentication module is enabled, it will ask for authentication no matter the username or password provided, even if the username is invalid. Scammers are definitely not the brightest individuals, but this could be a possible giveaway of a fake bank if they're looking out for it.

**Comment by ethanv2** [comment]
Copying in @SinisterSpatula, as that's his module.

Right now, modules work completely on the client side, so they have no way of knowing if the credentials entered are valid or not until the form has already been sent off - by which time it will be too late. I'm going to keep this open and, maybe, add it to a bumper list of things to do for the module API - which will allow things like this to take place by letting modules talk to the server for information.

If you can think of any other enhancements for modules that could be added using the new module API, please open another issue for any you think of.

**Comment by SinisterSpatula** [comment]
When I wrote it, there wasn't a good way that I could find to pass in the username/password FIRST and then do the auth.  I only saw a way to do the auth, and then submit name/pass.  Ideally, yes, this would be much better to do it after the name/pass have been verified.  I'm not sure this really should be a module anyway, it would probably be better as a built in function that itself becomes theme-able.

**Comment by ethanv2** [comment]
I definetely think this should remain a module. The DSJAS core is meant to be as minimal as possible as a kind of blank canvas for you to add extra things on top. Once the module API is written up, this will be 100% possible to do.

**Comment by SinisterSpatula** [comment]
If it stays a module, does that mean people will need a separate 2FA module for every theme they desire?  Or will it be possible to in a theme, re-skin this single module?  I'm still trying to wrap my head around the theme system.  I understand it's based on hooks, and I guess I'd have to add hooks to this module that themes can grab hold of?  Or are hooks part of the core and out of reach for a module.

**Comment by ethanv2** [comment]
Absolutely not. Modules are specifically designed to work regardless of the installed theme. This is why we use a hook system: the hooks don't care about the surrounding content or the style overrides the theme provides (which will affect the module too). All that needs to happen is that DSJAS loads the module at the location of the specified hook. As for who provides the hook locations, the core provides some of them and the theme provides the rest. Good examples of ones which the core provides are the "all" hook - which selects for everything - and the "login" hook - which selects for the login pages. The theme may provides ones such as "menubar" and "footer" to place content in specific places on the page. This way, as long as the theme provides "module descriptors" (basically just locations for hooks), modules work regardless of the theme.

**Comment by SinisterSpatula** [comment]
I guess I need to see an example/tutorial like "let's change the color of the spinner in 2FA from our theme...  We over ride the hook that lives in the login page like this, and then change the color by selecting the spinner div, etc.  Looking back at the code, I'm shocked I was able to get it working at all.  Really, I just followed your structure for the frustrating login process, and a bunch of trial and error.

**Comment by ethanv2** [comment]
Early fix in df84f65e1f

**Comment by SinisterSpatula** [comment]
Nice!  So it will do the 2FA after user authentication with that change?

**Comment by ethanv2** [comment]
I am now happy to mark this as done and ready to ship next release
